### PR TITLE
CSP-500 - AanHub Outer Api GET /Regions

### PR DESCRIPTION
**Pull request by LiamWright**
Initial endpoint for ApprenticeAan endpoint.

Regions controller/handler implemented as well as base classes needed for successful API calls.

ApprenticeAan also had not been added to the parent project, which has since been done now.

### CODE REVIEW

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Program.cs [hunk 1]
```diff
@@ -42,10 +44,6 @@
 
 // Configure the HTTP request pipeline.
 
-if (app.Environment.IsDevelopment())
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Program.cs:45]: Was this removed on purpose?

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQuery.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+﻿using MediatR;
+
+namespace SFA.DAS.ApprenticeAan.Application.Regions.Queries.GetRegions
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQuery.cs:3]: This should be along side the Query Handler.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/RegionsController.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            try
+            {
+                var result = await _mediator.Send(new GetRegionsQuery());
+
+                if (result == null) return NotFound();
+
+                return Ok(result);
+            }
+            catch { return BadRequest(); }
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/RegionsController.cs:33]: Remove try catch. Any errors should just bubble up and result in 500 response.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/RegionsController.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+        public RegionsController(IMediator mediator) => _mediator = mediator;
+
+        /// <summary>
+        ///     Get list of regions
+        /// </summary>
+        /// <returns></returns>
+        [HttpGet]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(string), StatusCodes.Status204NoContent)]
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/RegionsController.cs:21]: Remove 204 response type.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/RegionsController.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        [ProducesResponseType(typeof(GetRegionsQueryResult), StatusCodes.Status200OK)]
+        public async Task<IActionResult> GetRegions()
+        {
+            try
+            {
+                var result = await _mediator.Send(new GetRegionsQuery());
+
+                if (result == null) return NotFound();
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/RegionsController.cs:29]: I would return a 500 here as the expectations is to always get some data back. Also it is possible that inner API returned a 500.

**Review by DevenKShah — CHANGES_REQUESTED** [review]
We are trying to minimize the dependency on the shared projects hence I had the ApiClient bits in Api project. All these needs to be moved back please. There should be no changes to the shared projects.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Program.cs (hunk 1 above)
Comment (LiamWright) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Program.cs:45]: Yes, SONAR flagged this as a security hotspot

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQuery.cs (hunk 2 above)
Comment (LiamWright) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQuery.cs:3]: I'm not sure what you mean by this. `GetRegionsQuery` is in the same folder. Does it need to be in the same class?

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/RegionsController.cs (hunk 3 above)
Comment (LiamWright) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/RegionsController.cs]: Done

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/RegionsController.cs (hunk 4 above)
Comment (LiamWright) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/RegionsController.cs]: Done

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/RegionsController.cs (hunk 5 above)
Comment (LiamWright) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/RegionsController.cs]: Implemented

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Program.cs (hunk 1 above)
Comment (DevenKShah) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Program.cs:45]: That cannot be right, we do need to see exception page when debugging.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Common/Constants.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿namespace SFA.DAS.ApprenticeAan.Application.Common
+{
+    public static class Constants
+    {
+        public static class AanHubApiUrls
+        {
+            private const string ApiPrefix = "/api";
+            public const string GetRegionsUrl = $"{ApiPrefix}/regions";
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Common/Constants.cs:8]: Don't see a point of having specific urls as constants. This pattern will become complicated when you have dynamic routes to build.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQuery.cs (hunk 2 above)
Comment (DevenKShah) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQuery.cs:3]: duh moment.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQueryResult.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+﻿using SFA.DAS.ApprenticeAan.Application.Entities;
+
+namespace SFA.DAS.ApprenticeAan.Application.Regions.Queries.GetRegions
+{
+    public class GetRegionsQueryResult
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQueryResult.cs:5]: We dont have to have Result object if it does not make sense, like in this case the Response can easily be List<Region>

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application.UnitTests/Regions/Queries/GetRegions/GetRegionsQueryHandlerTests.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+            GetRegionsQuery query,
+            [Frozen] Mock<IAanHubApiClient<AanHubApiConfiguration>> apiClient,
+            [Frozen(Matching.ImplementedInterfaces)]
+            GetRegionsQueryHandler handler)
+        {
+            apiClient.Setup(x =>
+                    x.GetWithResponseCode<GetRegionsQueryResult>(It.IsAny<GetRegionsQueryRequest>()))
+                .ReturnsAsync(new ApiResponse<GetRegionsQueryResult>(null!, HttpStatusCode.BadRequest, string.Empty));
+
+            var result = await handler.Handle(query, CancellationToken.None);
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application.UnitTests/Regions/Queries/GetRegions/GetRegionsQueryHandlerTests.cs:47]: We need a way to communicate non success response further up so web consumer can react accordingly. This one I am not too bothered as we have static data.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Program.cs (hunk 1 above)
Comment (LiamWright) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Program.cs:45]: SONAR cloud doesn't keep the logs, so this historical run is the most detailed report I can get you, but it did in fact flag it.

image


Not sure what's changed with it now, but I've reinstated it.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQueryHandler.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        private readonly IAanHubApiClient<AanHubApiConfiguration> _apiClient;
+
+        public GetRegionsQueryHandler(IAanHubApiClient<AanHubApiConfiguration> apiClient) => _apiClient = apiClient;
+
+        public async Task<GetRegionsQueryResult?> Handle(GetRegionsQuery request, CancellationToken cancellationToken)
+        {
+            var regionsResult = await _apiClient.GetWithResponseCode<GetRegionsQueryResult>(new GetRegionsQueryRequest());
+
+            if (regionsResult.StatusCode != HttpStatusCode.OK || regionsResult.Body == null) return null;
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQueryHandler.cs:19]: Need to add a log error for unsuccessful response here as we do not expect this to not return data.

**Review by DevenKShah — APPROVED** [review]
Few more minor suggestions. I am approving as it all looks good otherwise.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQueryResult.cs (hunk 7 above)
Comment (LiamWright) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQueryResult.cs:5]: No, I thought that. But we might as well keep it mirrored to AanHub.Api if it's not too much overhead?

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQueryHandler.cs (hunk 9 above)
Comment (LiamWright) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQueryHandler.cs]: Added logError for unsuccessful calls

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Common/Constants.cs (hunk 6 above)
Comment (DevenKShah) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Common/Constants.cs:8]: IMHO it is just unnecessary hop in the code to check the URL. I am expecting that one request will have one endpoint to associate with, so dont think there is any benefit. If we find a benefit of doing this down the line then we can always refactor.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQueryResult.cs (hunk 7 above)
Comment (DevenKShah) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Regions/Queries/GetRegions/GetRegionsQueryResult.cs:5]: I am not too fussed, but I would rather avoid any boiler plate code.

**Review by DevenKShah — DISMISSED** [review]

**Review by MarkFCain — APPROVED** [review]

**Pull request merged**